bn-apps/cordapp-updates-distribution/design/design.md

```diff
@@ -0,0 +1,149 @@
+* Fetch metadata about a single or a range of versions from a remote Maven repository.
+* Download a single or a range of versions from a remote repository. `cds-lib` is packaging-type agnostic and can be used to pull down any type of artifacts from any Maven2 compatible repository. Support of classifiers will allow devs to publish CorDapps targeting different hardware / software configurations.
+
+`cds-lib` will support pluggable transports:
+* *HTTP(s)*. Available in Maven Resolver out-of-the-box with proxy- and repository- level authentications support.
+* *Corda Flows*. Transport implementation that allows transferring files over Corda Flows.
+* *Corda RPC*. Its essentially the same as *Corda Flows*, with the difference that the *RPC* version will be used if `cds-lib` is invoked from outside of a Corda node.
+
+`cds-lib` will be invokable from a command line, with command line parameters taking precedence over the configuration file. CLI invocation *could* look like:
+
+```
+# Will pull down a single version of cordapp and its dependencies
+> java -jar cds-lib.jar
+    -Dcordapp="com.my.company.name:corda-name:1.0"
+    -DremoteRepo="https://my-company-repo.com"
+    -DlocalRepo="/path/to/my/local/repo"
+    -Dtransport="rpc"
+    -Dconfiguration="config.properties"
+
+# Will pull down all versions of cordapps and their dependencies
+> java -jar cds-lib.jar
+    -Dcordapp="com.my.company.name:corda-name:[0,)"
+    -DremoteRepo="https://my-company-repo.com"
+    -DlocalRepo="/path/to/my/local/repo"
+    -Dtransport="http"
+    -Dconfiguration="config.properties"
+
+```
+
+Exact configuration parameters will be documented during the implementation.
+
+##### cds-cordapp
+
+`cds-cordapp` will provide the following functionality on *top* of the `cds-lib`:
+* Scheduled state for members to periodically sync their local repos with a single remote repository.
+* Flows for BNO to notify members about new CorDapp version availability. Members CorDapp will automatically attempt to download a new version.
+* Flows for BNO to notify members about CorDapp version revocations. If a version of CorDapp was revoked, members are expected to manually update their nodes to the last not-revoked version ASAP.
+* Flows for BNO to collect reports from members about CorDapp versions installed on their nodes. Only versions of CorDapps related to *this* Business Network should be reported. CDS will rely on the information provided in CorDapp `MANIFEST` files until a better API is available.
+* Flows for members to manually request lists of available / revoked versions and their descriptions from BNO.
```

Comment:
We should make sure all these flows are shell friendly, and perhaps just add a set of shell commands that act as aliases, to give a slightly more discoverable and enjoyable shell experience.

Reply:
Yeps, sounds good. Do we have something on the backlog to support custom shell extensions? Maybe API to allow cordapps to advertise their custom shell commands? Or otherwise we'd have to work from the assumption that CDS is installed on a node.

Reply:
Yeps, will a couple of lines about that